### ISSUE TITLE
Make `storage`, `prune`, and `check` optional

### ISSUE THREAD

**Issue by plasticrake** [jeffaco/duplicacy-util#32]
I have a use case to only run `copy` on a repository and never `backup`. However the `storage` section is required in the configuration file. I'd like to be able to safely run with `-a`, but I'd have to make sure to always only run with `-copy -check -prune`. 

I also have a use case to never run `prune` on certain repositories. The reason for skipping prune is that the docs say that [only one device should run prune per storage](https://forum.duplicacy.com/t/prune-command-details/1005), so I'd like to only have one machine running `prune`.

```
     backup, check        copy, check, prune (to offsite)
                                check, prune (local)
(local source devices) -> (LAN SFTP storage server) -> (offsite cloud storage)
```

Would you be open to this utility supporting this design via configuration files? Thank you!

**Comment by jeffaco** [comment]
This is a weird one for sure. I can see where it would come up if you use specific machines for specific backup tasks. I'm not entirely sure how you do this and keep sequencing of jobs intact between systems, but you must have figured this out. Or maybe not, because you know that if sequencing gets off, it will fix itself over time?

So, trying to understand your ask, you do recognize that `duplicacy-util` does support this today, right? The issue is that you want to run with `-a`, and `-a` specifically says to do all operations, but you don't want that. Is that your ask?

You mentioned a number of things here:

1. You need to specify things in the configuration that you may never do. My thoughts: Yeah, that's true. I'm unlikely to change this part, however, as I view this as a pretty edge case, and there is an easy workaround (specify what you want to do on the command line).

2. You want to use `-a` to just "do everything", but only have it do what's appropriate for the system. My thoughts: I'm unlikely to want to change `-a`, that's clearly documented for it's purpose. However, I am open to considering a new qualifier to run operations as defined in the backup configuration file.

I'm thinking perhaps something like this to fulfill your request:

* In the backup configuration file:

```
operations: ["copy", "check", "prune"]
```

* Add some new command line qualifier to run with those options (perhaps `-cops`, I think with a smile, to stand for configured options). If `-cops` (I'm open to a different name) was specified, we'd run with what was specified in the backup configuration file. If nothing was specified in the backup configuration file, we'd emit an error and exit.

I'm open to making `duplicacy-util` do what you want, but of course I don't want to change behavior for everyone. And since nearly everyone needs all sections specified in the configuration, I don't want to change the code to not require all sections (except copy). I don't want to change `-a` behavior, as that is clearly specified for what it does. But I'm fine adding a new qualifier to do what you want.

Please let me know your thoughts on this, thanks. I'm sure we can come to agreement on this. Also, I'm happy to take a PR for this, if you'd like, since you already submitted to the project, once we iron out specifics. I'm also fine doing it myself, it should only take 10 minutes or so to do the above (maybe 15 minutes with unit tests, which we should have).

**Comment by plasticrake** [comment]
My main concern would be accidentally triggering a backup when I only want a copy. On my SFTP server that is running the copy the dummy "source" repository is empty, and I believe it would mark all those items for deletion on the offsite storage. 

Currently with `-a` it will run a copy if the copy section is there, and will skip it if not. I don't see it as a huge change from the existing behavior to have the other sections not required and skip if they are not in the configuration (perhaps emit a warning, or another flag to enable this behavior). If the `-cops` as you propose would do that I'm fine switching from `-a` to `-cops`, just as long as `-cops` would work with those sections missing. If I'm required to have the `storage` section I'd be worried of accidentally running `-a` on a repository that would be destructive to run the `backup`.

To summarize I have  repository that is destructive to run `-a` as it is today, so I'd want either to have `-a` skip the backup if not in the configuration, or just have `-a` fail completely (since the section is missing). I'm fine with using `-copy` or `-cops`, as long as `-copy` can run without the `storage` configuration. Although I do see value in using `-a` or `-cops` that uses the configuration file to determine the operations to run. Otherwise I will need to keep that scripted separately for my repositories that differ instead of relying on the dulicacy-util configuration.

**Comment by jeffaco** [comment]
My thought here is that you would totally abandon use of `-a` all the time. You would set up your backup configuration files appropriately so, after that, you would **always** use `-cops`. You'd get finger memory to use `-cops` rather than `-a`. Your scripting would use `-cops` as well.

At that point, the backup configuration file (which contains all sections) decides what `-cops` actually does. You could even have the same backup config filename do different things based on the machine it's running on (just change the `operations:` line in the backup configuration file for that machine).

I like this because its not invasive to people using the software today, but flexible enough to give you what you want (if I understand what you want), even on a per-machine basis for the same repository (with a one-line change to the backup configuration file).

That's my thought, anyway.

**Comment by jeffaco** [comment]
@plasticrake You use Mac or Linux, right? Or do you use Windows?

I'm wondering if a better way is to use shell substitution to do what you want. So, to help avoid mistakes, say I add one thing to the backup configuration file: `disable_all_option: true`. This would make the `-a` option invalid for that backup (this would be there solely to protect against mistakes). Don't add `operations:` to the backup configuration. Instead, rely on shell substitution (you can do this on Windows too depending on your shell).

Example: In your .bashrc (or whatever login script you have, powershell supports this notion too on Windows), add something like:

DU_OPS="-backup -copy"

Then, when running `duplicacy-util`, you can do something like:

```
duplicacy-util -f mybackup $DU_OPS
```

Shell substitution would make this `duplicacy-util -f mybackup -backup -copy`. And off you go.

This is super easy for me, and seems to give you what you need. Let me know. Or, if you can get used to never using `-a`, perhaps I do nothing?

**Comment by jeffaco** [comment]
@plasticrake Still waiting for feedback ...

**Comment by plasticrake** [comment]
Hi @jeffaco, on mobile so will be brief. I’m using Mac and Linux. Little hard to follow all the different methods that have been proposed but ideally I’d want:

* if -a or -backup is used either error and do nothing or run only the operations that are desired (skip backup, etc)

* Don’t require the storage section in config or any section that is not relevant to the configuration

**Comment by plasticrake** [comment]
One other thing. I've built Ansible roles that will setup duplicacy and duplicacy-util. I've been polishing it up and finishing the documentation and hope to have on github soon. But it will do a complete install, configuration, and setup with cron / launchd with just configuring via Ansible YML files. It generates the YML config files for duplicacy-util as well. It would be easier to setup all the cron jobs to just use `-a` or another common parameter (like `-cops` discussed above) and do the configuration via the file instead of having to have have custom command line options per backup repository, but if that's how you'd like to have it done I can work around that.

**Comment by jeffaco** [comment]
@plasticrake As for as I'm aware, yours is the only use case where you don't need a storage section, and yours is a pretty unusual use case. Thus, I'm not included to change how this works. I'd rather you just include the storage section and have it not be used.

See above messages. I'm inclined to not add `-cops` after all, as you could trivially get the same thing with environment variable substitution, which is supported on all platforms. Again, see above messages for my rational.

As for as `-a`, this is really that you got "used to" this and developed finger memory. In my case, I have some repositories where I routinely use `-a`, and some where I never do. On repositories where `duplicacy check` is [very slow](https://github.com/gilbertchen/duplicacy/issues/397), I can't use `-a` (On those repositories, I only run `duplicacy check` a few times/week instead of daily). But this is all done via `launchd` scripting on Mac OS/X, so after setup I never really worry about that anymore.

I think that, given you can use shell substitution to deal different default qualifiers, I'm not sure that anything is really needed. `duplicacy-util` has the flexibility to do what you want; you just need to tell it what you want.

As for Ansible, I'd probably set it up to use `-a` by default since that's what the vast majority of the people will use. But you can control that with Ansible configuration file to say what you want to normally run.

**Comment by Algebro7** [comment]
Sorry to resurrect an old issue, but I just ran into this ~~and find the current options really counter-intuitive. I'm having trouble thinking of a use-case for why `-a`  attempt to run operations that haven't been defined in the `.yaml`, rather than just running all operations defined in it.~~

For my use-case, I have a couple of machines that only want to `backup` and `check`, but I'm still getting errors complaining that prune locations haven't been defined, even when using `-backup` and `-check` instead of `-a`, which also doesn't make sense based on my understanding of the explanations above. 

Do you have any suggestions on how to configure `duplicacy-util` to let me only run a backup and check? My understanding based on the readme was that you could just run `-backup` and `-check` but that doesn't seem to be the case.

**Comment by Algebro7** [comment]
Actually, looking a bit more at the source code, it looks like a `prune` configuration is required, even if `cmdPrune` is false:

```
configBackup.go

	if len(config.pruneInfo) == 0 {
		err = errors.New("no prune locations defined in configuration")
		logError(nil, fmt.Sprint("Error: ", err))
	}
```

`check` does the same thing, although it looks like it might only be an issue when reading the configs, as `backup-ops.go` appears to honor the `cmd*` arguments when deciding which operations to perform.

I've hacked in a fix but it's probably too ugly for you to want a PR, although I can provide it if you aren't going to be able to get around to fixing it. Just let me know.
